tracpro/contacts/fields.py

```diff
@@ -41,14 +44,75 @@ def render(self, name, value, attrs=None):
         return mark_safe(''.join(output))
 
 
+def validate_twitter_handle(handle):
+    handle_regex = r'^[a-zA-Z0-9_]{1,15}$'
+    if not re.match(handle_regex, handle):
+        raise ValidationError(
+            _("That is not a valid format Twitter handle.  A valid handle has only "
+              "letters A-Z, numbers 0-9, and underscores (_), and has up to 15 "
+              "characters."))
+
+
+def validate_numeric_twitter_id(numeric_id):
+    if not numeric_id.isdigit():  # True only if all chars are digits
+        raise ValidationError(
+            _("That is not a valid format Twitter ID.  A valid numeric ID has only "
+              "numbers 0-9."))
+
+
+def validate_twitter_id(twitter_id):
+    if '#' in twitter_id:
+        numeric_id, handle = twitter_id.split('#', 1)
+        validate_twitter_handle(handle)
+        validate_numeric_twitter_id(numeric_id)
+    else:
+        raise ValidationError(
+            _("That is not a valid format Twitter ID.  The correct format is: "
+              "<numeric id>#<Twitter handle>"))
+
+
+def validate_phone_used(number):
+    from .models import Contact
+    contacts = Contact.objects.all()
+    phone_numbers = set()
+    for contact in contacts:
+        if contact.urn.startswith('tel'):
+            phone_numbers.add(contact.urn.split(':')[1])
+
+    if number in phone_numbers:
+        raise ValidationError(
+             _("This phone number is already being used by another contact."))
+
+
+def validate_phone(number):
+    try:
+        parsed = phonenumbers.parse(number)
+    except Exception as error:
```

Comment:
Here, we should only catch the specific exception that we're expecting if the number is not valid. If anything goes wrong that we didn't expect, we'd like to find out about it and not hide it.  This case isn't quite so bad in that way, since it's still going to report an error, but not all exceptions have an ``_msg`` attribute, and it's just a good habit to get into to catch the narrowest possible set of exceptions.